docs: rewrite structural directives (#40015)
angular/angular: aio/content/guide/built-in-directives.md

</div>

{@a ngcontainer}

> ## `<ng-container>`

Reviewer: I would like to have a conversation about our style guide when we have code in a heading. I am concerned that this type of heading isn't "skimmable." Sure, if I already know about `<ng-container>`, I'm good to go! But what if I don't?

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

> Angular's `<ng-container>` is a grouping element that doesn't render anything in the DOM.
> This means you can use `<ng-container>` to apply directives without affecting page layout or styling.

Reviewer: A suggestion: Put what user's needs first. For example:

```suggestion
Sometimes, you might want to apply a directive without affecting the page layout or styling. For these situations, you can use Angular's `<ng-container>` element.
```

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

> Here's a conditional paragraph using `<ng-container>`.

Reviewer: ```suggestion
The following code uses `<ng-container>` to conditionally display a clause within a sentence.
```
I worry that the use of "here" seems too abrupt. I'd also like to verify that it's localizable.

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (ngif-ngcontainer)" region="ngif-ngcontainer"></code-example>

<div class="lightbox">
  <img src='generated/images/guide/structural-directives/good-paragraph.png' alt="ngcontainer paragraph with proper style">
</div>

> To conditionally exclude an `<option>`, use `<ng-container>`.

Reviewer: ```suggestion
You can also use `<ng-container> to conditionally exclude options in a list.
```

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (select-ngcontainer)" region="select-ngcontainer"></code-example>

<div class="lightbox">
  <img src='generated/images/guide/structural-directives/select-ngcontainer-anim.gif' alt="ngcontainer options work properly">
</div>

<div class="alert is-helpful">

> Import the `ngModel` directive from `FormsModule` and add `FormsModule` in the imports: [...] section of the Angular module metadata, in which you want to use it.

Reviewer: I'm not sure why this note is here?

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

</div>

`<ng-container>` is a syntax element that Angular parses like the curly braces in a JavaScript `if`-block:

<code-example language="javascript">
  if (someCondition) {
    statement1;
    statement2;
    statement3;
  }

</code-example>

Without those braces, JavaScript would only execute the first statement when you intend to conditionally execute all of them as a single block.

> The `<ng-container>` satisfies a similar need in Angular templates.

Reviewer: This is another example of going into "why" when I don't think we need to. I do like line 421, but I think it should be up at the beginning of the section. I don't think the code example and lines 432 and 433 are necessary, because we are assuming that the reader has some understanding of JavaScript.

Author: Per offline conversation, resolving these comments as this work is addressed in https://github.com/angular/angular/pull/39816

</div>

{@a ngcontainer}
## `<ng-container>`

The Angular `<ng-container>` is a grouping element that doesn't interfere with styles or layout because Angular doesn't put it in the DOM.

You can use [`<ng-container>`](guide/built-in-directives#ngcontainer) when there's no single element to host the directive.

Here's a conditional paragraph using `<ng-container>`.

<code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (ngif-ngcontainer)" region="ngif-ngcontainer"></code-example>

<div class="lightbox">
  <img src='generated/images/guide/structural-directives/good-paragraph.png' alt="ngcontainer paragraph with proper style">
</div>

To conditionally exclude an `<option>`, use `<ng-container>`.

> <code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (select-ngcontainer)" region="select-ngcontainer"></code-example>

Author: What do you think about me doing it in #39816 after this (#40015) lands?

> <code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (select-ngcontainer)" region="select-ngcontainer"></code-example>

Reviewer 2: Thats reasonable

> <code-example path="structural-directives/src/app/app.component.html" header="src/app/app.component.html (select-ngcontainer)" region="select-ngcontainer"></code-example>

Reviewer: @jelbourn don't forget that our primary focused is on _how_ to do things. Conceptual content (the _why_) belongs elsewhere. And even then, sometimes the best way to explain why is to create content (like tutorials) that address specific use cases.

The _why_ question can get very tiresome for users. While it's useful once, it's burdensome to read through on subsequent visits.
